src/gs-background-tile.c

```diff
@@ -0,0 +1,292 @@
+		gtk_widget_hide (tile->price);
+		gtk_widget_show (tile->installed_icon);
+		return;
+	}
+
+	price = gs_app_get_price (tile->app);
+	if (price == NULL || (gs_price_get_amount (price) == (gdouble) 0)) {
+		/* TRANSLATORS: This is the price of the app, as in gratis */
+		price_str = g_utf8_strup (_("Free"), -1);
+		gtk_label_set_label (GTK_LABEL (tile->price), price_str);
+	} else {
+		price_str = gs_price_to_string (price);
+		gtk_label_set_label (GTK_LABEL (tile->price), price_str);
+	}
+
+	gtk_widget_show (tile->price);
+	gtk_widget_hide (tile->installed_icon);
+}
+
+static void
+update_tile_info (GsBackgroundTile *tile)
+{
+	g_autofree gchar *name = NULL;
+	AtkObject *accessible = gtk_widget_get_accessible (GTK_WIDGET (tile));
+	const gchar *summary = NULL;
+
+	if (gs_app_is_installed (tile->app)) {
+		/* TRANSLATORS: This is the name and state of an app for the ATK object */
+		name = g_strdup_printf (_("%s (Installed)"), gs_app_get_name (tile->app));
+	} else if (gs_app_get_state (tile->app) == AS_APP_STATE_AVAILABLE) {
+		name = g_strdup (gs_app_get_name (tile->app));
+	}
+
+	summary = gs_app_get_summary (tile->app);
+
+	if (GTK_IS_ACCESSIBLE (accessible)) {
+		atk_object_set_name (accessible, name);
+		atk_object_set_description (accessible, summary);
+	}
+
+	gtk_label_set_label (GTK_LABEL (tile->name_label), gs_app_get_name (tile->app));
+	gtk_label_set_label (GTK_LABEL (tile->summary_label), summary);
+
+	update_tile_price (tile);
+}
+
+static gboolean
+app_state_changed_idle (gpointer user_data)
+{
+	GsBackgroundTile *tile = GS_BACKGROUND_TILE (user_data);
+
+	update_tile_info (tile);
+	g_object_unref (tile);
+
+	return G_SOURCE_REMOVE;
+}
+
+static void
+app_state_changed (GsApp *app,
+		   GParamSpec *pspec,
+		   GsBackgroundTile *tile)
+{
+	/* we call the function in idle because the state-change
+	 * notification could be coming from a different thread */
+	g_idle_add (app_state_changed_idle, g_object_ref (tile));
+}
+
+static void
+app_background_tile_css_changed (GsApp *app,
+				 const char *metadata,
+				 GsBackgroundTile *tile)
+{
+	update_tile_background (tile);
+}
+
+static void
+gs_background_tile_set_app (GsAppTile *app_tile,
+			    GsApp *app)
+{
+	GsBackgroundTile *tile = GS_BACKGROUND_TILE (app_tile);
+	GdkPixbuf *pixbuf = NULL;
+
+	g_return_if_fail (GS_IS_APP (app) || app == NULL);
+
+	if (tile->app)
+		g_signal_handlers_disconnect_by_func (tile->app, app_state_changed, tile);
+
+	g_set_object (&tile->app, app);
+
+	if (app == NULL)
+		return;
+
+	gtk_stack_set_visible_child_name (GTK_STACK (tile->stack), "content");
+
+	g_signal_connect (tile->app, "notify::state",
+			  G_CALLBACK (app_state_changed), tile);
+	g_signal_connect (tile->app, "notify::name",
+			  G_CALLBACK (app_state_changed), tile);
+	g_signal_connect (tile->app, "notify::summary",
+			  G_CALLBACK (app_state_changed), tile);
+	g_signal_connect (tile->app, "notify::key-colors",
+			  G_CALLBACK (app_key_colors_changed), tile);
```

Comment:
You could use `g_signal_connect_swapped()` and `call update_tile_background()` directly from this and the next callback.